File: src/1019_Next_Greater_Node_In_Linked_List.py
# 1019_Next_Greater_Node_In_Linked_List
# https://leetcode.cn/problems/next-greater-node-in-linked-list/


# Definition for singly-linked list.
# 超时
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        nums = []
        stack = []
        
        cur = head
        while cur:
            nums.append(cur.val)
            cur = cur.next
        n = len(nums)
        
        ans = [0] * n
        
        for i in range(n):
            while stack and nums[i] > nums[stack[-1]]:
                idx = stack.pop()
                ans[idx] = nums[i]
            stack.append(i)
            
        return ans 
    
        
        
# 超时    
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        res = []
        
        def nextLargeNode(node: ListNode):
            if node is None:
                return
            
            curr = node.next
            ans = node.val
            while curr:
                if curr.val > ans:
                    ans = curr.val
                    break
                curr = curr.next
            
            nonlocal res
            if ans > node.val:
                res.append(ans)
            else:
                res.append(0)
                
        while head:
            nextLargeNode(head)
            head = head.next
        
        return res

File: src/1019_Next_Greater_Node_In_Linked_List.rs
// https://leetcode.cn/problems/next-greater-node-in-linked-list/submissions/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}


impl Solution {
    pub fn next_larger_nodes(head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut ans = vec![];
        let mut stack: Vec<(i32, usize)> = vec![];
        let mut cur = 0;
        let mut head = head.as_ref();

        while let Some(h) = head {
            ans.push(0);

            while !stack.is_empty() && stack[stack.len() - 1].0 < h.val {
                ans[stack[stack.len() - 1].1] = h.val;
                stack.pop();
            }

            stack.push((h.val, cur));
            cur += 1;
            head = h.next.as_ref();
        }

        ans
    }
}

impl Solution {
    pub fn next_larger_nodes(mut head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut ans = vec![];
        let mut stack: Vec<(i32, usize)> = vec![];
        let mut cur = 0;

        while let Some(mut node) = head {
            ans.push(0);

            while !stack.is_empty() && stack[stack.len() - 1].0 < node.val {
                ans[stack[stack.len() - 1].1] = node.val;
                stack.pop();
            }

            stack.push((node.val, cur));
            cur += 1;
            head = node.next.take();
        }

        ans
    }
}